docs/glossary.rst

```diff
@@ -0,0 +1,274 @@
+
+    resource
+        Describes an API hook for an interaction mediated by a Context. Data flow
+        is described as *immutable* resources (generally produced as Operation outputs)
+        that can be consumed by binding to Operation inputs or by extracting as *results*
+        from the API. Some interactions cannot be represented in terms of producers
+        and subscribers of immutable data events: *Mutable* resources cannot be
+        managed by the Context as data events and require different work scheduling
+        policies that either (a) allows arbitrary (unscheduled) call-back through the API framework,
+        (b) dispatch the mutable resource collaboration to another Context, or (c)
+        allow operations to bind and interact with an interface not specified by the
+        API or not known to the responsible Context implementation. Examples include
+        the Context-provided *ensemble_reduce* functionality, the ensemble simulation
+        signaling facility (by which extension code can terminate a simulation early),
+        and the binding mechanism by which MD extension code can be attached to an
+        *MD* operation as a plugin. The nature of a resource is indicated by the
+        namespace of its *port* in the work record.
+
+    run time
+        .. todo:: Define the scope to be conveyed by the noun "run time".
+
+    simulation segment
+    trajectory segment
+        A sequence of molecular simulation iterations or frames produced
+        deterministically (within numerical limits) under well-determined
+        parameters. For the purposes of discussing checkpoint intervals or the
+        minimum amount of work executed between API calls, it is useful to
+        distinguish between full simulation trajectories and the irreducible
+        unit of work supported by a simulation library. In the simplest API use
+        cases, a simulation library does not interact with the API during
+        production of a simulation segment, and allows for reinitialization
+        between simulation segments. This allows for unambiguous labeling of the
+        artifacts of a segment. Optimizations may focus on reducing overhead
+        between successive simulation segments (minimizing reinitialization).
+        Extensions may introduce abstractions for well-characterized non-constant
+        parameters, such as time-varying lambda values, though such abstractions
+        are not required in the API since the effect can be achieved through
+        binding to a mutable resource (with details beyond the scope of the API)
+        owned by another operation whose state and action is well characterized
+        for the segment.
+
+    task
+        The smallest divisible unit of work representable within the API.
+        Generally an :term:`operation` or a portion of an operation that has
+        been decomposed in terms of decoupled parallelism or execution checkpoint
+        interval.
+
+    work
+    graph
```

Comment:
@mrshirts comments:
> I'm not excited about the term "graph" since it has heavy connotations as an abstraction or representation of a thing, rather than a thing itself. "The graph" is problematic, since one could have many graphs (which process or subprocess does it correspond to". I could be ok with "the process graph', as long as it is modified.

"graph" is definitely an abstraction here. There is no Graph data structure, and one can indeed have many graphs. There is no "process graph." At the lowest levels, there is potentially a "task graph," and generally there is a data flow graph.

We should be more careful about how we use the term, and additional qualification is a good idea. But we need to clarify some other terms and develop terminology for several concepts if we want to remove any of these terms from our lexicon.
* What is the network of information and algorithms expressed by a user, and what is the hierarchy of conceptual and logical components?
* What are the individual units and collections of units of dispatchable, schedulable computation at the various API layers or levels of abstraction?